Handle focusability for plugin elements which has browsing context

Ahmad-S792 opened the pull request:
#### 61a5480546df3e43c1b258c65d12e9396b1b5f0b
<pre>
Handle focusability for plugin elements which has browsing context

<a href="https://bugs.webkit.org/show_bug.cgi?id=259420">https://bugs.webkit.org/show_bug.cgi?id=259420</a>

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/a2c82afad86cf4f85f91df76f858434b15fe6b13">https://chromium.googlesource.com/chromium/src.git/+/a2c82afad86cf4f85f91df76f858434b15fe6b13</a>

If `object` or `embed` has HTML/SVG contents,
focus navigation within documents under these
elements should work. This has been working fine
for `iframe`&apos;s.

The FocusController code checks for FrameOwnerElement,
which is parent of `iframe`, `object`, and `embed` if
it is isKeyboardFocusable() and has contentFrame().

HTMLPluginElement::isKeyboardFocsuable() didn&apos;t handle
the cases with HTML/SVG contents, only considered plugins.

* Source/WebCore/html/HTMLPlugInElement.cpp:
(HTMLPlugInElement::isKeyboardFocusable): As above for focusability
* Source/WebCore/html/HTMLFrameOwnerElement.h: Move &apos;isKeyboardFocusable&apos; from private to protected
* LayoutTests/fast/dom/focus-navigation-in-plugin.html: Add Test Case
* LayoutTests/fast/dom/resources/plugin-focus-subframe.html: Add Test Case resource
* LayoutTests/fast/dom/focus-navigation-in-plugin-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Added test to skip on iOS since it does not support &apos;keyDown&apos;

Canonical link: <a href="https://commits.webkit.org/266286@main">https://commits.webkit.org/266286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8336fb1593d85db83a86c69975fe05176eb17dcb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13392 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15129 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12747 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13471 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13758 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15429 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14210 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15806 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12086 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19134 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15465 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12756 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11951 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3276 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16350 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12598 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLFrameOwnerElement.h:
```diff
@@ -66,14 +66,15 @@ class HTMLFrameOwnerElement : public HTMLElement {
     virtual bool shouldLoadFrameLazily() { return false; }
     virtual bool isLazyLoadObserverActive() const { return false; }
 
+    bool isKeyboardFocusable(KeyboardEvent*) const override;
```
This could be under `protected:`

Ahmad-S792 replied:
Sure, will move. I am expecting this to fail on 'iOS-wk2' due to keyboard event, so I will add 'Platform Specific expectation' and this part in one go. :-)

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
Handle focusability for plugin elements which has browsing context

https://bugs.webkit.org/show_bug.cgi?id=259420

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/a2c82afad86cf4f85f91df76f858434b15fe6b13

If `object` or `embed` has HTML/SVG contents,
focus navigation within documents under these
elements should work. This has been working fine
for `iframe`'s.

The FocusController code checks for FrameOwnerElement,
which is parent of `iframe`, `object`, and `embed` if
it is isKeyboardFocusable() and has contentFrame().

HTMLPluginElement::isKeyboardFocsuable() didn't handle
the cases with HTML/SVG contents, only considered plugins.

* Source/WebCore/html/HTMLPlugInElement.cpp:
(HTMLPlugInElement::isKeyboardFocusable): As above for focusability
* Source/WebCore/html/HTMLFrameOwnerElement.h: Move 'isKeyboardFocusable' from private to protected
* LayoutTests/fast/dom/focus-navigation-in-plugin.html: Add Test Case
* LayoutTests/fast/dom/resources/plugin-focus-subframe.html: Add Test Case resource
* LayoutTests/fast/dom/focus-navigation-in-plugin-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Added test to skip on iOS since it does not support 'keyDown'

Canonical link: https://commits.webkit.org/266286@main
